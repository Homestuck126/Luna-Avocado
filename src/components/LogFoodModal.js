import React, { useState, useEffect } from "react";
import {
  View,
  Text,
  StyleSheet,
  Pressable,
  Modal,
  TextInput,
} from "react-native";

//----------------------------------------------------------------------------------}

const LogFoodModal = ({ isVisible, onClose, addFood }) => {
  const [modalVisible, setModalVisible] = useState(false);
  const [foodName, setFoodName] = useState("");
  const [calories, setCalories] = useState(0);
  const [protein, setProtein] = useState(0);
  const [carbohydrates, setCarbohydrates] = useState(0);
  const [fat, setFat] = useState(0);

  useEffect(() => {
    setModalVisible(isVisible);
  }, [isVisible]);

  const closeModal = () => {
    setModalVisible(false);
    onClose(false);
  };

  const handleSubmit = () => {
    //console.log({ foodName, calories, protein, carbohydrates, fat });
    if (calories > 0 && protein > 0 && carbohydrates > 0 && fat > 0) {
      addFood({
        id: null,
        name: foodName,
        macros: [
          Number(calories),
          Number(protein),
          Number(carbohydrates),
          Number(fat),
        ],
      });
    }
    setFoodName("");
    setCalories(0);
    setProtein(0);
    setCarbohydrates(0);
    setFat(0);
    closeModal();
  };

  return (
    <Modal
      animationType="slide"
      transparent={true}
      visible={modalVisible}
      onRequestClose={closeModal}
    >
      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          <TextInput
            style={styles.input}
            placeholder="Name of the food"
            value={foodName}
            onChangeText={setFoodName}
            autoCapitalize="none"
          />
          <TextInput
            style={styles.input}
            placeholder="Calories"
            value={calories}
            onChangeText={setCalories}
            keyboardType="numeric"
          />
          <TextInput
            style={styles.input}
            placeholder="Protein (g)"
            value={protein}
            onChangeText={setProtein}
            keyboardType="numeric"
          />
          <TextInput
            style={styles.input}
            placeholder="Carbohydrates (g)"
            value={carbohydrates}
            onChangeText={setCarbohydrates}
            keyboardType="numeric"
          />
          <TextInput
            style={styles.input}
            placeholder="Fat (g)"
            value={fat}
            onChangeText={setFat}
            keyboardType="numeric"
          />
          <View style={styles.buttonContainer}>
            <Pressable
              style={({ pressed }) => [
                styles.button,
                styles.buttonClose,
                {
                  backgroundColor: pressed
                    ? "rgba(52,152,219,0.7)"
                    : "rgb(52,152,219)",
                },
              ]}
              onPress={handleSubmit}
            >
              <Text style={styles.modalText}>Submit</Text>
            </Pressable>
            <Pressable
              style={({ pressed }) => [
                styles.button,
                styles.buttonClose,
                {
                  backgroundColor: pressed
                    ? "rgba(52,152,219,0.7)"
                    : "rgb(52,152,219)",
                },
              ]}
              onPress={() => {
                closeModal();
              }}
            >
              <Text style={styles.modalText}>Cancel</Text>
            </Pressable>
          </View>
        </View>
      </View>
    </Modal>
  );
};

// Reconstruct the object with `foodName` first
const orderedLogData = {
  foodName: console.log.foodName,
  calories: console.log.calories,
  protein: console.log.protein,
  carbohydrates: console.log.carbohydrates,
  fat: console.log.fat,
};

console.log(orderedLogData);

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 22,
  },
  modalView: {
    height: "60%", // Adjusted for additional content
    width: "80%",
    margin: 20,
    backgroundColor: "white",
    borderRadius: 20,
    padding: 35,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  input: {
    height: 40,
    width: "90%",
    marginBottom: 12,
    borderWidth: 1,
    padding: 10,
    borderRadius: 5,
    borderColor: "#ddd",
  },
  button: {
    padding: 10,
    borderRadius: 8,
    width: "45%",
    alignItems: "center",
    marginHorizontal: 5,
  },
  buttonContainer: {
    justifyContent: "center",
    flexDirection: "row",
  },
  modalText: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
  },
});

export default LogFoodModal;
