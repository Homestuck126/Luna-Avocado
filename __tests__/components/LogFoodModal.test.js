import { render, screen, fireEvent } from "@testing-library/react-native";
import LogFoodModal from "../../src/components/LogFoodModal";
import "@testing-library/jest-native";

describe("Log Food Modal", () => {
  it("calls onClose prop when Submit pressable is pressed", () => {
    const mockOnPress = jest.fn();

    const { getByTestId } = render(
      <LogFoodModal
        isVisible={true}
        onClose={mockOnPress}
        addFood={() => {}}
      ></LogFoodModal>
    );
    const submitButton = getByTestId("LogFoodModal:SubmitPressable:Press");
    fireEvent.press(submitButton);

    expect(mockOnPress).toHaveBeenCalled();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("calls onClose prop when Cancel pressable is pressed", () => {
    const mockOnPress = jest.fn();

    const { getByTestId } = render(
      <LogFoodModal
        isVisible={true}
        onClose={mockOnPress}
        addFood={() => {}}
      ></LogFoodModal>
    );
    const submitButton = getByTestId("LogFoodModal:CancelPressable:Press");
    fireEvent.press(submitButton);

    expect(mockOnPress).toHaveBeenCalled();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("displays input name and macros when TextInput changes", () => {
    render(
      <LogFoodModal
        isVisible={true}
        onClose={() => {}}
        addFood={() => {}}
      ></LogFoodModal>
    );

    const nameTextInput = screen.getByTestId(
      "LogFoodModal:NameTextInput:Change"
    );
    const CalorieTextInput = screen.getByTestId(
      "LogFoodModal:CalorieTextInput:Change"
    );
    const ProteinTextInput = screen.getByTestId(
      "LogFoodModal:ProteinTextInput:Change"
    );
    const CarbTextInput = screen.getByTestId(
      "LogFoodModal:CarbTextInput:Change"
    );
    const FatTextInput = screen.getByTestId("LogFoodModal:FatTextInput:Change");

    fireEvent.changeText(nameTextInput, "test");
    fireEvent.changeText(CalorieTextInput, "2000");
    fireEvent.changeText(ProteinTextInput, "100");
    fireEvent.changeText(CarbTextInput, "200");
    fireEvent.changeText(FatTextInput, "300");

    expect(nameTextInput).toHaveDisplayValue("test");
    expect(CalorieTextInput).toHaveDisplayValue("2000");
    expect(ProteinTextInput).toHaveDisplayValue("100");
    expect(CarbTextInput).toHaveDisplayValue("200");
    expect(FatTextInput).toHaveDisplayValue("300");
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
