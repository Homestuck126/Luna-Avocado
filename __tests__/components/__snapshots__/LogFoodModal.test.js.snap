// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Food Modal calls onClose prop when Cancel pressable is pressed 1`] = `null`;

exports[`Log Food Modal calls onClose prop when Submit pressable is pressed 1`] = `null`;

exports[`Log Food Modal displays input name and macros when TextInput changes 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 22,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "height": "60%",
          "margin": 20,
          "padding": 35,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
          "width": "80%",
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        onChangeText={[Function]}
        placeholder="Name of the food"
        style={
          {
            "borderColor": "#ddd",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 12,
            "padding": 10,
            "width": "90%",
          }
        }
        testID="LogFoodModal:NameTextInput:Change"
        value="test"
      />
      <TextInput
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Calories"
        style={
          {
            "borderColor": "#ddd",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 12,
            "padding": 10,
            "width": "90%",
          }
        }
        testID="LogFoodModal:CalorieTextInput:Change"
        value="2000"
      />
      <TextInput
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Protein (g)"
        style={
          {
            "borderColor": "#ddd",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 12,
            "padding": 10,
            "width": "90%",
          }
        }
        testID="LogFoodModal:ProteinTextInput:Change"
        value="100"
      />
      <TextInput
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Carbohydrates (g)"
        style={
          {
            "borderColor": "#ddd",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 12,
            "padding": 10,
            "width": "90%",
          }
        }
        testID="LogFoodModal:CarbTextInput:Change"
        value="200"
      />
      <TextInput
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Fat (g)"
        style={
          {
            "borderColor": "#ddd",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "marginBottom": 12,
            "padding": 10,
            "width": "90%",
          }
        }
        testID="LogFoodModal:FatTextInput:Change"
        value="300"
      />
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 8,
                "marginHorizontal": 5,
                "padding": 10,
                "width": "45%",
              },
              undefined,
              {
                "backgroundColor": "rgb(52,152,219)",
              },
            ]
          }
          testID="LogFoodModal:SubmitPressable:Press"
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Submit
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 8,
                "marginHorizontal": 5,
                "padding": 10,
                "width": "45%",
              },
              undefined,
              {
                "backgroundColor": "rgb(52,152,219)",
              },
            ]
          }
          testID="LogFoodModal:CancelPressable:Press"
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Cancel
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
